lightning/src/ln/onion_message.rs

```diff
@@ -0,0 +1,96 @@
+	pub encrypted_payload: Vec<u8>,
+}
+
+/// Onion messages can be sent and received to blinded routes, which serve to hide the identity of
+/// the recipient.
+pub struct BlindedRoute {
+	/// To send to a blinded route, the sender first finds a route to the unblinded
+	/// `introduction_node_id`, which can unblind its [`encrypted_payload`] to find out the onion
+	/// message's next hop and forward it along.
+	///
+	/// [`encrypted_payload`]: BlindedNode::encrypted_payload
+	pub introduction_node_id: PublicKey,
+	/// Creators of blinded routes supply the introduction node id's `blinding_point`, which the
+	/// introduction node will use in decrypting its [`encrypted_payload`] to forward the onion
+	/// message.
+	///
+	/// [`encrypted_payload`]: BlindedNode::encrypted_payload
+	pub blinding_point: PublicKey,
+	/// The blinded hops of the blinded route.
+	pub blinded_hops: Vec<BlindedNode>,
+}
+
+impl BlindedRoute {
+	/// Create a blinded route to be forwarded along `hops`.
+	pub fn new<Signer: Sign, K: Deref>(hops: Vec<PublicKey>, keys_manager: K) -> Result<Self, secp256k1::Error>
+		where K::Target: KeysInterface<Signer = Signer> {}
+}
+
+/// The final destination of an onion message.
+pub enum Destination {
+	/// We're sending this onion message to a node.
+	Node(PublicKey),
+	/// We're sending this onion message to a blinded route.
+	BlindedRoute(BlindedRoute),
+}
+
+/// A sender, receiver and forwarder of onion messages. In upcoming releases, this object will be
+/// used to retrieve invoices and fulfill invoice requests from offers.
+pub struct OnionMessager<Signer: Sign, K: Deref>
+where K::Target: KeysInterface<Signer = Signer>
+{
+	keys_manager: K,
+	// Coming soon:
+	// invoice_handler: InvoiceHandler,
+	// custom_handler: CustomHandler, // handles custom onion messages
+	pending_msg_events: Vec<MessageSendEvent>,
+}
+
+// Coming soon:
+// /// `InvoiceHandler` handles incoming invoice requests, invoices, and invoice errors.
+// pub trait InvoiceHandler {
+//   /// Handles an incoming invoice request.
+//   fn handle_invoice_request(invoice_request: InvoiceRequest) -> Result<Invoice, ()>;
+//   // Handles an incoming invoice.
+//   fn handle_invoice(invoice: Invoice) -> Result<(), ()>;
+//   // Handles an incoming invoice error.
+//   fn handle_invoice_error(invoice_error: InvoiceError);
+//   // Sends a payment.
+//   fn send_payment(..);
+// }
+
+impl<Signer: Sign, K: Deref> OnionMessager<Signer, K>
+	where K::Target: KeysInterface<Signer = Signer>,
+{
+	/// Constructs a new `OnionMessager` to send, forward, and delegate received onion messages to
+	/// their respective handlers.
+	pub fn new(keys_manager: K) -> Self {}
+
+	/// Send an empty onion message to `recipient`, routing it through `intermediate_nodes`.
+	// NOTE: sending custom TLVs and reply paths aren't included atm
+	pub fn send_onion_message(&self, recipient: Destination, intermediate_nodes: Vec<PublicKey>) -> Result<(), APIError> {}
```

Comment:
I suppose routing would play a role in determining `intermediate_nodes`. Would it be a single-path route and do you need more than the `PublicKey` of each hop?

Reply:
Yeah, single-path route. Only the `PublicKey` is needed!